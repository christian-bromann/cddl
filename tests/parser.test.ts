import fs from 'node:fs/promises'
import url from 'node:url'
import path from 'node:path'
import { describe, it, expect } from 'vitest'

import Lexer from '../src/lexer.js'
import Parser from '../src/parser.js'

const __dirname = url.fileURLToPath(new URL('.', import.meta.url))

describe('parser', () => {
    const testCases: { name: string, fixture: string }[] = [
        { name: 'should correctly parse CDDL file', fixture: 'example.cddl' },
        { name: 'can parse compositions', fixture: 'compositions.cddl' },
        { name: 'can parse ranges', fixture: 'ranges.cddl' },
        { name: 'can parse occurrences', fixture: 'occurrences.cddl' },
        { name: 'can parse arrays', fixture: 'arrays.cddl' },
        { name: 'can parse unwrapped arrays', fixture: 'unwrapping.cddl' },
        { name: 'can parse comments', fixture: 'comments.cddl' },
        { name: 'can parse choices', fixture: 'choices.cddl' },
        { name: 'can parse nested groups', fixture: 'nested.cddl' }
    ]

    for (const { name, fixture } of testCases) {
        it(name, async () => {
            const input = await fs.readFile(
                path.join(__dirname, '__fixtures__', fixture),
                'utf-8'
            )

            const l = new Lexer(input)
            const p = new Parser(l)
            expect(p.parse()).toMatchSnapshot()
        })
    }
})