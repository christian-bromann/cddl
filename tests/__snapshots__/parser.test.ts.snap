// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parser > can parse arrays 1`] = `
[
  {
    "Name": "unlimited-people",
    "Type": "array",
    "Values": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "person",
          },
        ],
      },
    ],
  },
  {
    "Name": "at-least-one-person",
    "Type": "array",
    "Values": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": Infinity,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "person",
          },
        ],
      },
    ],
  },
  {
    "Content": "some",
    "Type": "comment",
  },
  {
    "Name": "zero-or-one-person",
    "Type": "array",
    "Values": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "person",
          },
        ],
      },
    ],
  },
  {
    "Content": "comment",
    "Type": "comment",
  },
  {
    "Name": "one-or-two-people",
    "Type": "array",
    "Values": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 2,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "person",
          },
        ],
      },
    ],
  },
  {
    "Name": "zero-or-two-people",
    "Type": "array",
    "Values": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 2,
          "n": 0,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "person",
          },
        ],
      },
    ],
  },
  {
    "Name": "two-or-infinity-people",
    "Type": "array",
    "Values": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": Infinity,
          "n": 2,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "person",
          },
        ],
      },
    ],
  },
  {
    "IsChoiceAddition": false,
    "Name": "person",
    "Properties": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "name",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "age",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "uint",
        ],
      },
    ],
    "Type": "group",
  },
]
`;

exports[`parser > can parse choices 1`] = `
[
  {
    "IsChoiceAddition": false,
    "Name": "attire",
    "PropertyType": [
      {
        "Type": "literal",
        "Unwrapped": false,
        "Value": "bow tie",
      },
      {
        "Type": "literal",
        "Unwrapped": false,
        "Value": "necktie",
      },
      {
        "Type": "literal",
        "Unwrapped": false,
        "Value": "Internet attire",
      },
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": false,
    "Name": "protocol",
    "PropertyType": [
      {
        "Type": "literal",
        "Unwrapped": false,
        "Value": 6,
      },
      {
        "Type": "literal",
        "Unwrapped": false,
        "Value": 17,
      },
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": false,
    "Name": "address",
    "Properties": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "delivery",
          },
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "delivery",
    "Properties": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "street",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "number",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "uint",
        ],
      },
      [
        {
          "Comment": "",
          "HasCut": false,
          "Name": "",
          "Occurrence": {
            "m": 1,
            "n": 1,
          },
          "Type": {
            "Type": "group",
            "Unwrapped": false,
            "Value": "city",
          },
        },
        {
          "Comment": "",
          "HasCut": true,
          "Name": "po-box",
          "Occurrence": {
            "m": 1,
            "n": 1,
          },
          "Type": [
            "uint",
          ],
        },
      ],
      [
        {
          "Comment": "",
          "HasCut": false,
          "Name": "",
          "Occurrence": {
            "m": 1,
            "n": 1,
          },
          "Type": {
            "Type": "group",
            "Unwrapped": false,
            "Value": "city",
          },
        },
        {
          "Comment": "",
          "HasCut": true,
          "Name": "foo",
          "Occurrence": {
            "m": 1,
            "n": 1,
          },
          "Type": [
            "tstr",
            "uint",
          ],
        },
        {
          "Comment": "",
          "HasCut": false,
          "Name": "",
          "Occurrence": {
            "m": 1,
            "n": 1,
          },
          "Type": {
            "Type": "group",
            "Unwrapped": false,
            "Value": "foobar",
          },
        },
        {
          "Comment": "",
          "HasCut": true,
          "Name": "per-pickup",
          "Occurrence": {
            "m": 1,
            "n": 1,
          },
          "Type": [
            {
              "Type": "literal",
              "Unwrapped": false,
              "Value": true,
            },
          ],
        },
      ],
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "city",
    "Properties": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "name",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "zip-code",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "uint",
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": true,
    "Name": "attire",
    "PropertyType": [
      {
        "Type": "literal",
        "Unwrapped": false,
        "Value": "swimwear",
      },
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": true,
    "Name": "delivery",
    "Properties": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "lat",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "float",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "long",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "float",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "drone-type",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
    ],
    "Type": "group",
  },
]
`;

exports[`parser > can parse comments 1`] = `
[
  {
    "Content": "This is a comment",
    "Type": "comment",
  },
  {
    "IsChoiceAddition": false,
    "Name": "person",
    "Properties": [
      {
        "Comment": "a good employer",
        "HasCut": true,
        "Name": "employer",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
    ],
    "Type": "group",
  },
  {
    "Content": "foobar",
    "Type": "comment",
  },
  {
    "Content": "This is another",
    "Type": "comment",
  },
  {
    "Content": "very nice comment",
    "Type": "comment",
  },
  {
    "IsChoiceAddition": false,
    "Name": "foo",
    "Properties": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "bar",
          },
        ],
      },
    ],
    "Type": "group",
  },
  {
    "Content": "some comment",
    "Type": "comment",
  },
  {
    "Content": "another comment",
    "Type": "comment",
  },
  {
    "IsChoiceAddition": false,
    "Name": "som",
    "Properties": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "g",
          },
        ],
      },
    ],
    "Type": "group",
  },
  {
    "Content": "some g",
    "Type": "comment",
  },
]
`;

exports[`parser > can parse compositions 1`] = `
[
  {
    "IsChoiceAddition": false,
    "Name": "person",
    "Properties": [
      {
        "Comment": "an identity",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "identity",
          },
        ],
      },
      {
        "Comment": "some employer",
        "HasCut": true,
        "Name": "employer",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "dog",
    "Properties": [
      {
        "Comment": "an identity again",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "identity",
          },
        ],
      },
      {
        "Comment": "some length",
        "HasCut": true,
        "Name": "leash-length",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "float",
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "identity",
    "Properties": [
      {
        "Comment": "an age",
        "HasCut": true,
        "Name": "age",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "a name",
        "HasCut": true,
        "Name": "name",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
    ],
    "Type": "group",
  },
  {
    "Name": "Geography",
    "Type": "array",
    "Values": [
      {
        "Comment": "a city",
        "HasCut": true,
        "Name": "city",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "some coordinates",
        "HasCut": true,
        "Name": "gpsCoordinates",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "GpsCoordinates",
          },
        ],
      },
    ],
  },
  {
    "IsChoiceAddition": false,
    "Name": "GpsCoordinates",
    "Properties": [
      {
        "Comment": "degrees, scaled by 10^7",
        "HasCut": true,
        "Name": "longitude",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "uint",
        ],
      },
      {
        "Comment": "degreed, scaled by 10^7",
        "HasCut": true,
        "Name": "latitude",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "uint",
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "located-samples",
    "Properties": [
      {
        "Comment": "some",
        "HasCut": false,
        "Name": "sample-point",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "comments",
        "HasCut": false,
        "Name": "samples",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Name": "",
            "Type": "array",
            "Values": [
              {
                "Comment": "",
                "HasCut": false,
                "Name": "",
                "Occurrence": {
                  "m": Infinity,
                  "n": 1,
                },
                "Type": "float",
              },
            ],
          },
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "extensible-map-example",
    "Properties": [
      {
        "Comment": "some",
        "HasCut": true,
        "Name": "optional-key",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "really",
        "HasCut": true,
        "Name": "optional-key2",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "interesting",
        "HasCut": true,
        "Name": "optional-key3",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "comments",
        "HasCut": false,
        "Name": "tstr",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "any",
          },
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "personWithoutComment",
    "Properties": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "identity",
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "employer",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "dogWithoutComment",
    "Properties": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "identity",
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "leash-length",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "float",
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "identityWithoutComment",
    "Properties": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "age",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "name",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
    ],
    "Type": "group",
  },
  {
    "Name": "GeographyWithoutComment",
    "Type": "array",
    "Values": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "city",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "gpsCoordinates",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "GpsCoordinates",
          },
        ],
      },
    ],
  },
  {
    "IsChoiceAddition": false,
    "Name": "GpsCoordinatesWithoutComment",
    "Properties": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "longitude",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "uint",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "latitude",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "uint",
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "located-samples-without-comments",
    "Properties": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "sample-point",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "",
        "HasCut": false,
        "Name": "samples",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Name": "",
            "Type": "array",
            "Values": [
              {
                "Comment": "",
                "HasCut": false,
                "Name": "",
                "Occurrence": {
                  "m": Infinity,
                  "n": 1,
                },
                "Type": "float",
              },
            ],
          },
        ],
      },
    ],
    "Type": "group",
  },
  {
    "IsChoiceAddition": false,
    "Name": "extensible-map-example-without-comments",
    "Properties": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "optional-key",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "optional-key2",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "optional-key3",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "",
        "HasCut": false,
        "Name": "tstr",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "any",
          },
        ],
      },
    ],
    "Type": "group",
  },
]
`;

exports[`parser > can parse nested groups 1`] = `
[
  {
    "IsChoiceAddition": false,
    "Name": "f",
    "Properties": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "o",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "IsChoiceAddition": false,
            "Name": "",
            "Properties": [
              {
                "Comment": "",
                "HasCut": true,
                "Name": "o",
                "Occurrence": {
                  "m": 1,
                  "n": 1,
                },
                "Type": [
                  {
                    "IsChoiceAddition": false,
                    "Name": "",
                    "Properties": [
                      {
                        "Comment": "",
                        "HasCut": true,
                        "Name": "b",
                        "Occurrence": {
                          "m": 1,
                          "n": 1,
                        },
                        "Type": [
                          {
                            "IsChoiceAddition": false,
                            "Name": "",
                            "Properties": [
                              {
                                "Comment": "",
                                "HasCut": true,
                                "Name": "a",
                                "Occurrence": {
                                  "m": 1,
                                  "n": 1,
                                },
                                "Type": [
                                  {
                                    "IsChoiceAddition": false,
                                    "Name": "",
                                    "Properties": [
                                      {
                                        "Comment": "",
                                        "HasCut": true,
                                        "Name": "r",
                                        "Occurrence": {
                                          "m": 1,
                                          "n": 1,
                                        },
                                        "Type": [
                                          "text",
                                        ],
                                      },
                                    ],
                                    "Type": "group",
                                  },
                                ],
                              },
                            ],
                            "Type": "group",
                          },
                        ],
                      },
                    ],
                    "Type": "group",
                  },
                ],
              },
            ],
            "Type": "group",
          },
        ],
      },
    ],
    "Type": "group",
  },
]
`;

exports[`parser > can parse occurrences 1`] = `
[
  {
    "IsChoiceAddition": false,
    "Name": "someGroup",
    "Properties": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "optional",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "anotherOptional",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "atLeastOne",
        "Occurrence": {
          "m": Infinity,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "numberedOccurence",
        "Occurrence": {
          "m": 42,
          "n": 23,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "withoutLeftSide",
        "Occurrence": {
          "m": 42,
          "n": 0,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "withoutRightSide",
        "Occurrence": {
          "m": Infinity,
          "n": 23,
        },
        "Type": [
          "tstr",
        ],
      },
    ],
    "Type": "group",
  },
]
`;

exports[`parser > can parse ranges 1`] = `
[
  {
    "IsChoiceAddition": false,
    "Name": "device-address",
    "PropertyType": [
      {
        "Type": "group",
        "Unwrapped": false,
        "Value": "byte",
      },
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": false,
    "Name": "max-byte",
    "PropertyType": [
      {
        "Type": "literal",
        "Unwrapped": false,
        "Value": 255,
      },
      "float",
      "int",
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": false,
    "Name": "my_uri",
    "PropertyType": [
      {
        "Type": "tag",
        "Unwrapped": false,
        "Value": {
          "NumericPart": 6.32,
          "TypePart": "tstr",
        },
      },
      "tstr",
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": false,
    "Name": "byte",
    "PropertyType": [
      {
        "Type": "range",
        "Unwrapped": false,
        "Value": {
          "Inclusive": true,
          "Max": {
            "Type": "group",
            "Unwrapped": false,
            "Value": "max-byte",
          },
          "Min": 0,
        },
      },
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": false,
    "Name": "first-non-byte",
    "PropertyType": [
      {
        "Type": "literal",
        "Unwrapped": false,
        "Value": 256,
      },
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": false,
    "Name": "byte1",
    "PropertyType": [
      {
        "Type": "range",
        "Unwrapped": false,
        "Value": {
          "Inclusive": false,
          "Max": {
            "Type": "group",
            "Unwrapped": false,
            "Value": "first-non-byte",
          },
          "Min": 0,
        },
      },
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": false,
    "Name": "min..max",
    "PropertyType": [
      {
        "Type": "literal",
        "Unwrapped": false,
        "Value": 0,
      },
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": false,
    "Name": "byte2",
    "PropertyType": [
      {
        "Type": "group",
        "Unwrapped": false,
        "Value": "min..max...first-non-byte",
      },
    ],
    "Type": "variable",
  },
  {
    "IsChoiceAddition": false,
    "Name": "byte3",
    "PropertyType": [
      {
        "Type": "range",
        "Unwrapped": false,
        "Value": {
          "Inclusive": false,
          "Max": {
            "Type": "group",
            "Unwrapped": false,
            "Value": "first-non-byte",
          },
          "Min": "min..max",
        },
      },
    ],
    "Type": "variable",
  },
]
`;

exports[`parser > can parse unwrapped arrays 1`] = `
[
  {
    "Name": "basic-header",
    "Type": "array",
    "Values": [
      {
        "Comment": "",
        "HasCut": true,
        "Name": "field1",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "int",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "field2",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "text",
        ],
      },
    ],
  },
  {
    "Name": "advanced-header",
    "Type": "array",
    "Values": [
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": true,
            "Value": "basic-header",
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "field3",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "bytes",
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "field4",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": true,
            "Value": "time",
          },
        ],
      },
    ],
  },
]
`;

exports[`parser > should correctly parse CDDL file 1`] = `
[
  {
    "IsChoiceAddition": false,
    "Name": "Example",
    "Properties": [
      {
        "Comment": "summary of test",
        "HasCut": true,
        "Name": "title",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "longer description of test",
        "HasCut": true,
        "Name": "description",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "tstr",
        ],
      },
      {
        "Comment": "Is this a success or failure test",
        "HasCut": true,
        "Name": "fail",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          "bool",
        ],
      },
      {
        "Comment": "Inputs to create the test",
        "HasCut": true,
        "Name": "input",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "Inputs",
          },
        ],
      },
      {
        "Comment": "Intermediate values for debugging",
        "HasCut": true,
        "Name": "intermediates",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "Intermediates",
          },
        ],
      },
      {
        "Comment": "Outputs of the test",
        "HasCut": true,
        "Name": "output",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "Outputs",
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "attire",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "literal",
            "Unwrapped": false,
            "Value": "bow tie",
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "someBool",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "literal",
            "Unwrapped": false,
            "Value": true,
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "someOtherBool",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "literal",
            "Unwrapped": false,
            "Value": false,
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "count",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "literal",
            "Unwrapped": false,
            "Value": 123,
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "rating",
        "Occurrence": {
          "m": Infinity,
          "n": 0,
        },
        "Type": [
          {
            "Type": "literal",
            "Unwrapped": false,
            "Value": 0.5724646875815566,
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "hexa",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "literal",
            "Unwrapped": false,
            "Value": "0x123",
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "bin",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "literal",
            "Unwrapped": false,
            "Value": "0b123",
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": false,
        "Name": "",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "group",
            "Unwrapped": false,
            "Value": "foobar",
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "samples",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Name": "",
            "Type": "array",
            "Values": [
              {
                "Comment": "",
                "HasCut": false,
                "Name": "",
                "Occurrence": {
                  "m": Infinity,
                  "n": 1,
                },
                "Type": "float",
              },
            ],
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "multitype",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          "float",
          "tstr",
          {
            "Type": "literal",
            "Unwrapped": false,
            "Value": "foobar",
          },
        ],
      },
      {
        "Comment": "",
        "HasCut": true,
        "Name": "my_uri",
        "Occurrence": {
          "m": 1,
          "n": 1,
        },
        "Type": [
          {
            "Type": "tag",
            "Unwrapped": false,
            "Value": {
              "NumericPart": 6.32,
              "TypePart": "tstr",
            },
          },
          "tstr",
        ],
      },
    ],
    "Type": "group",
  },
]
`;
